fluss-common/src/main/java/org/apache/fluss/row/BinaryMap.java

```diff
@@ -0,0 +1,166 @@
+ * </pre>
+ *
+ * <p>Influenced by Apache Spark UnsafeMapData.
+ *
+ * @since 0.9
+ */
+@PublicEvolving
+public class BinaryMap extends BinarySection implements InternalMap {
+
+    private static final long serialVersionUID = 1L;
+
+    private transient BinaryArray keys;
+    private transient BinaryArray values;
+
+    private final transient Supplier<BinaryArray> keyArraySupplier;
+    private final transient Supplier<BinaryArray> valueArraySupplier;
+    private final transient Supplier<BinaryMap> nestedMapSupplier;
+
+    public BinaryMap() {
+        this(
+                (Supplier<BinaryArray> & Serializable) PrimitiveBinaryArray::new,
+                (Supplier<BinaryArray> & Serializable) PrimitiveBinaryArray::new,
+                (Supplier<BinaryMap> & Serializable) BinaryMap::new);
+    }
+
+    public BinaryMap(
+            Supplier<BinaryArray> keyArraySupplier,
+            Supplier<BinaryArray> valueArraySupplier,
+            Supplier<BinaryMap> nestedMapSupplier) {
+        this.keyArraySupplier = keyArraySupplier;
+        this.valueArraySupplier = valueArraySupplier;
+        this.nestedMapSupplier = nestedMapSupplier;
+    }
+
+    @Override
+    public int size() {
+        return keys.size();
+    }
+
+    @Override
+    public void pointTo(MemorySegment[] segments, int offset, int sizeInBytes) {
+        // Read the numBytes of key array from the first 4 bytes.
+        final int keyArrayBytes = BinarySegmentUtils.getInt(segments, offset);
+        assert keyArrayBytes >= 0 : "keyArraySize (" + keyArrayBytes + ") should >= 0";
+        final int valueArrayBytes = sizeInBytes - keyArrayBytes - 4;
+        assert valueArrayBytes >= 0 : "valueArraySize (" + valueArrayBytes + ") should >= 0";
+
+        // see BinarySection.readObject, on this call stack, keys and values are not initialized
+        if (keys == null) {
+            keys = createKeyArrayInstance();
+        }
+        keys.pointTo(segments, offset + 4, keyArrayBytes);
+        if (values == null) {
+            values = createValueArrayInstance();
+        }
+        values.pointTo(segments, offset + 4 + keyArrayBytes, valueArrayBytes);
+
+        assert keys.size() == values.size();
+
+        this.segments = segments;
+        this.offset = offset;
+        this.sizeInBytes = sizeInBytes;
+    }
+
+    /** Creates a {@link BinaryArray} instance for keys with the nested data type information. */
+    protected BinaryArray createKeyArrayInstance() {
+        return keyArraySupplier.get();
+    }
+
+    /** Creates a {@link BinaryArray} instance for values with the nested data type information. */
+    protected BinaryArray createValueArrayInstance() {
+        return valueArraySupplier.get();
+    }
+
+    /** Creates a nested {@link BinaryMap} with the nested data type information. */
+    protected BinaryMap createNestedMapInstance() {
+        return nestedMapSupplier.get();
+    }
```

Comment:
Follow the pattern used in `BinaryArray` by declaring these as **abstract methods** and implementing them in subclasses.  

This approach eliminates the need for non-serializable fields like `keyArraySupplier`, which become `null` after deserialization and can lead to `NullPointerException` here at runtime.